### PR TITLE
add some probes and update reducers to return { result }.

### PR DESCRIPTION
**Pull request by santiago-elustondo**
1) add some probes (these are by and large not new. only new one is "root zone")
------
2) update reducers to return { result }.

TLDR: changed the abstract `Reducer.deriveShallowState()` method to expect as a return value an object with a `result` property. Other properties in this object may be used by the reducer to maintain private state.

The state of a reducer is often compared to the previous state using strict equality, be it by scanners or by higher-order reducers, to determine whether the lastEvent triggered a "change of state".

Sometimes reducers would like to collect data from a particular event, but not trigger a change of state. A change of state may have reducer-specific semantic implications that are not warranted, and would negatively impact separation of concerns and optimization. Instead, the reducer might would to hold onto this data, and release it during the next state change.

For example: the LastElapsedCD reducer returns a new state every time that the last elapsed change detection cycle changes. It would like to return as part of this state a list of all components change detected during this CD. The lastElapsedCD object would look something like:
```javascript
{
    startEID: ...
    startPerfStamp: ... 
    finishPerfStamp: ...
    componentsChecked: [
       {
           componentType: ...
           startPerfStamp: ...
           finishPerfStamp: ...
       },
       {
           componentType: ...
           startPerfStamp: ...
           finishPerfStamp: ...
       },
    ]
}
```

Now, each component check will arrive during a different event, and we dont want to release a new state every time this happens. We want to wait until we have everything, and release all at once.

This PR updates the reducer API so that the `Reducer.deriveShallowState()` abstract method is expected to return an object, which will be the `ShallowState`. The property `result` in `ShallowState` is what will be used by scanners or higher-order reducers to know if a meaningful change of state has occurred.

### CODE REVIEW

**Comment by santiago-elustondo** [comment]
took a while to do the rebasing, but i got it :)

File: packages/tools/tslint.config.json [hunk 1]
```diff
@@ -1,7 +1,21 @@
...
   "defaultSeverity": "error",
-  "extends": ["tslint:recommended", "tslint-config-prettier"],
+  "extends": [
+    "tslint:recommended",
+    "tslint-config-prettier"
+  ],
   "jsRules": {},
-  "rules": {},
+  "rules": {
+    "interface-name": false,
```
Comment (andrewthauer) [inline packages/tools/tslint.config.json]: I'm not a fan of overriding most of these rules. Most of these are actually good practice to use. In the odd case where something needs to be exempted they should be tslint-disabled inline (ideally with a comment note about why). At the very least in the specific module and not cart blanche across all packages.

File: packages/core/src/framework/reducers/reducer.ts [hunk 2]
```diff
@@ -4,14 +4,20 @@ import { objToPairs, merge } from '../utils'
...
-  prevState
-  prevDepState
+  prevShallowState
+  prevDepResults
   // @todo: can we do this better?
   resetDependency: (depName: string) => void
 }
 
+// @todo: types
+export interface ShallowState {
```
Comment (andrewthauer) [inline packages/core/src/framework/reducers/reducer.ts]: I would say types that are exported should be in their own file. Ideally one exported interface, class, etc. within a file.

Comment (andrewthauer) [inline packages/core/src/framework/reducers/reducer.ts]: Can we have a JSDoc style comment to describe what this interface / shape is for. It's definitely not clear by the name alone. Or perhaps consider a more meaningful name?

File: packages/core/src/framework/reducers/reducer.ts [hunk 3]
```diff
@@ -4,14 +4,20 @@ import { objToPairs, merge } from '../utils'
...
+  result?: any,
```
Comment (andrewthauer) [inline packages/core/src/framework/reducers/reducer.ts]: Can this be typed to `ShallowState<T>` where result is `result?: T`?

File: packages/core/src/framework/reducers/reducer.ts [hunk 4]
```diff
@@ -4,14 +4,20 @@ import { objToPairs, merge } from '../utils'
...
+  auxiliary?: { [key: string]: any }
```
Comment (andrewthauer) [inline packages/core/src/framework/reducers/reducer.ts]: Naming here seems a bit vague. Auxiliary could mean literal anything. Can we come up with something more clear?

File: packages/core/src/reducers/current-cycle.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  deriveShallowState({ prevShallowState = INIT_STATE, nextEvent, nextDepResults, prevDepResults }) {
+    const { isStable: nextIsStable, currentTask: nextTask } = nextDepResults
+    const { isStable: prevIsStable } = prevDepResults
+
+    // @definition: if we just became unstable, then we've entered a new cycle, thus started a new `job`
+    if ((prevIsStable || prevIsStable === undefined) && !nextIsStable)
+      return {
+        result: {
+          startEID: nextEvent.id,
```
Comment (andrewthauer) [inline packages/core/src/reducers/current-cycle.ts:26]: Can we avoid the acronyms and be explicit? For instance calling it `startEventId` would be more appropriate.

File: packages/core/src/reducers/current-cycle.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import { AuguryEvent } from '../framework/events'
+import { Reducer } from '../framework/reducers'
+
+import { CurrentNgZoneTaskReducer } from './current-ng-zone-task'
+import { IsStableReducer } from './is-stable'
+
+const INIT_STATE = {
+  result: undefined
```
Comment (andrewthauer) [inline packages/core/src/reducers/current-cycle.ts:8]: Might be better to use `null` instead of undefined?

File: packages/core/src/reducers/current-cycle.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { AuguryEvent } from '../framework/events'
+import { Reducer } from '../framework/reducers'
+
+import { CurrentNgZoneTaskReducer } from './current-ng-zone-task'
+import { IsStableReducer } from './is-stable'
+
+const INIT_STATE = {
```
Comment (andrewthauer) [inline packages/core/src/reducers/current-cycle.ts:7]: What do you think about being explicit and calling these `INITIAL_STATE` vs `INIT_STATE`? I think it's a bit more clear as init could in theory imply `initialization` instead of `initial`. `DEFAULT_STATE` would also work imo.

File: packages/core/src/framework/reducers/reducer.ts [hunk 8]
```diff
@@ -35,22 +54,20 @@ export abstract class Reducer {
 
     const nextShallowState = this.deriveShallowState({
```
Comment (andrewthauer) [inline packages/core/src/framework/reducers/reducer.ts:60]: As discussed, I think it might be worth considering if there is a way to avoid tracking child state and parent state in order to potentially simplify reconciling. If possible it would be nice give as much responsibility to the child reducers and somehow avoid introducing this concepts of prevPepState, nextDepState. If this the root objective for passing these through can be avoided, or centralized so child reducers do not need to know about the concept this serves it could make it easier to reason about the problem a bit. Anyways, my 2 conceptual cents :)

**Review by andrewthauer — CHANGES_REQUESTED** [review]
As I mentioned it's probably easier to talk about some of these things in person. I've left some notes to think about. It also looks like you may need to run prettier across the board. Most files don't seem to be following the ruleset defined.
